### PR TITLE
EventHub stream provider logging/monitoring

### PR DESCRIPTION
**Pull request by jason-bragg**
This addresses log/metrics logging infrastructure in stream provider as well as some high priority cases.  More error logging will likely be needed.  

Added logging and metrics from eventhub receiver.
Added logging from eventhub message cache.

### CODE REVIEW

File: src/OrleansProviders/Streams/Common/PooledCache/PooledQueueCache.cs [hunk 1]
```diff
@@ -65,14 +68,15 @@ public class PooledQueueCache<TQueueMessage, TCachedMessage>
         /// For ordered reliable queues we shouldn't need to notify on every purged event, only on the last event 
         ///   of every set of events that get purged.
         /// </summary>
-        public Action<TCachedMessage> OnPurged { get; set; }
+        public Action<TCachedMessage?,TCachedMessage?> OnPurged { get; set; }
```
Comment (jdom) [inline src/OrleansProviders/Streams/Common/PooledCache/PooledQueueCache.cs]: maybe an update to the comments to understand what are the 2 different messages?

File: src/OrleansServiceBus/ErrorCode.cs [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+    /// Orleans ServiceBus error codes
+    /// </summary>
+    internal enum OrleansServiceBusErrorCode
+    {
+        /// <summary>
+        /// Start of orlean servicebus errocodes
+        /// </summary>
+        ServiceBus = 1<<16,
+
+        FailedPartitionRead = ServiceBus,
```
Comment (jdom) [inline src/OrleansServiceBus/ErrorCode.cs]: Not sure if it's an actual convention or not, but I noticed that everywhere where we have error code constants, we skip the feature base code (meaning that this should have been `FailedPartitionRead = ServiceBus + 1`)

File: src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs [hunk 3]
```diff
@@ -45,16 +48,32 @@ protected EventHubQueueCache(int defaultMaxAddCount, IStreamQueueCheckpointer<st
         /// <returns></returns>
         protected abstract bool TryCalculateCachePressureContribution(StreamSequenceToken token, out double cachePressureContribution);
 
-        public void Dispose()
+        void IDisposable.Dispose()
```
Comment (jdom) [inline src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs]: I would disadvise implementing IDisposable explicitly... what prompted this change?

File: src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs [hunk 4]
```diff
@@ -115,14 +134,30 @@ public bool IsUnderPressure()
...
+        /// Handles cache purge signals
+        /// </summary>
+        /// <param name="lastItemPurged"></param>
+        /// <param name="newestItem"></param>
+        protected override void OnPurge(CachedEventHubMessage? lastItemPurged, CachedEventHubMessage? newestItem)
         {
+            if (lastItemPurged.HasValue && newestItem.HasValue)
+            {
+                log.Info($"CachePeriod: EnqueueTimeUtc: {LogFormatter.PrintDate(lastItemPurged.Value.EnqueueTimeUtc)} to {LogFormatter.PrintDate(newestItem.Value.EnqueueTimeUtc)}, DequeueTimeUtc: {LogFormatter.PrintDate(lastItemPurged.Value.DequeueTimeUtc)} to {LogFormatter.PrintDate(newestItem.Value.DequeueTimeUtc)}");
+            }
```
Comment (jdom) [inline src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs]: I'd check log.IsInformational() first at least, as this looks like a relatively costly string concatenation

**Comment by jason-bragg** [comment]
Addressed comments, rebased, squashed.

File: src/OrleansProviders/Streams/Common/PooledCache/PooledQueueCache.cs [hunk 5]
```diff
@@ -33,6 +34,8 @@ public class PooledQueueCache<TQueueMessage, TCachedMessage>
         private readonly CachedMessagePool<TQueueMessage, TCachedMessage> pool;
         private readonly ICacheDataAdapter<TQueueMessage, TCachedMessage> cacheDataAdapter;
         private readonly ICacheDataComparer<TCachedMessage> comparer;
+        private readonly Logger logger;
+        private int itemCount;
```
Comment (jdom) [inline src/OrleansProviders/Streams/Common/PooledCache/PooledQueueCache.cs]: I imagine int is enough, but just wanted to confirm with you, maybe there's scenarios when we would want long?
Also, all access to this is without Interlocked, I assume this doesn't need to be thread safe?

File: src/OrleansServiceBus/Providers/Streams/EventHub/EventHubAdapterReceiver.cs [hunk 6]
```diff
@@ -35,22 +48,35 @@ internal class EventHubAdapterReceiver : IQueueAdapterReceiver, IQueueCache
...
+            baseLogger = logger;
+            this.logger = logger.GetSubLogger("-receiver");
             config = partitionConfig;
+
+            hubReceiveTimeMetric = $"Orleans.ServiceBus.EventHub.ReceiveTime_{config.Hub.Path}";
+            partitionReceiveTimeMetric = $"Orleans.ServiceBus.EventHub.ReceiveTime_{config.Hub.Path}-{config.Partition}";
+            hubReadFailure = $"Orleans.ServiceBus.EventHub.ReadFailure_{config.Hub.Path}";
+            partitionReadFailure = $"Orleans.ServiceBus.EventHub.ReadFailure_{config.Hub.Path}-{config.Partition}";
+            hubMessagesRecieved = $"Orleans.ServiceBus.EventHub.MessagesReceived_{config.Hub.Path}";
+            partitionMessagesRecieved = $"Orleans.ServiceBus.EventHub.MessagesReceived_{config.Hub.Path}-{config.Partition}";
```
Comment (jdom) [inline src/OrleansServiceBus/Providers/Streams/EventHub/EventHubAdapterReceiver.cs]: nit: typo in Received (in the field name only)

File: src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs [hunk 7]
```diff
@@ -13,28 +14,47 @@ namespace Orleans.ServiceBus.Providers
...
         private readonly AveragingCachePressureMonitor cachePressureMonitor;
 
+        /// <summary>
+        /// Logic used to store queue position
+        /// </summary>
         protected IStreamQueueCheckpointer<string> Checkpointer { get; }
 
-        protected EventHubQueueCache(int defaultMaxAddCount, IStreamQueueCheckpointer<string> checkpointer, ICacheDataAdapter<EventData, TCachedMessage> cacheDataAdapter, ICacheDataComparer<TCachedMessage> comparer)
+        /// <summary>
+        /// Construct ecent hub queue cache.
```
Comment (jdom) [inline src/OrleansServiceBus/Providers/Streams/EventHub/EventHubQueueCache.cs]: nit: typo

File: src/OrleansProviders/Streams/Common/PooledCache/PooledQueueCache.cs [hunk 8]
```diff
@@ -61,18 +64,19 @@ public class PooledQueueCache<TQueueMessage, TCachedMessage>
         }
 
         /// <summary>
-        /// Called with the last item puraged after a cache purge has run.
+        /// Called with the newst item in the cache and last item puraged after a cache purge has run.
```
Comment (jdom) [inline src/OrleansProviders/Streams/Common/PooledCache/PooledQueueCache.cs]: nit: typo in "newst" and "puraged"

**Comment by jdom** [comment]
LGTM :+1: 
In the last review I had only minor comments related to typos. If they are easy to just fix and update, please do, otherwise I can just merge

**Pull request merged**